### PR TITLE
feat: implement text transformer for xapian searches.

### PR DESCRIPTION
**Pull request by Alexanderlacuna**
PR implements text transformer for xapian queries to feed to fahamu api.

### CODE REVIEW

File: gn2/wqflask/views.py [hunk 1]
```diff
@@ -268,11 +268,47 @@ def gsearchtable():
...
+        "rif",
+        "wiki",
+    }
+    xapian_operators = {"AND", "NOT", "OR", "XOR", "NEAR", "ADJ"}
+    range_prefixes = {"mean", "peak", "position", "peakmb", "additive", "year"}
+    query_context = ["genes"]
+    cleaned_query_parts = []
+    for token in query.split():
+        if token in xapian_operators or any(
+            prefix in token for prefix in range_prefixes if ".." in token
```
Comment (jnduli) [inline gn2/wqflask/views.py]: What conditions are we handling here? I can only think of `mean:8..` and `Hs:chr4:9930021`. These can be handled by:
```
if prefix in range_prefixes:
  continue
```
I'm not sure the `".." in token` is needed

File: gn2/wqflask/views.py [hunk 2]
```diff
@@ -268,11 +268,47 @@ def gsearchtable():
...
+            continue
+        prefix, _, suffix = token.partition(":")
+        if prefix in xapian_prefixes:
+            query_context.insert(0, prefix)
+            cleaned_query_parts.append(f"{prefix} {suffix}")
+        else:
+            cleaned_query_parts.append(prefix)
+    cleaned_query = " ".join(cleaned_query_parts)
+    context = ",".join(query_context)
+    return f"Provide answer on {cleaned_query} context {context}"
```
Comment (jnduli) [inline gn2/wqflask/views.py]: Question: 
1. Does this results in better results and should we somehow integrate this in the normal searches?
2. Should we add some type of formatting to the text i.e.
```bash
# if I search for:
rif:glioma species:human group:gtex_v8
# we pass in the query:
Provide answer on glioma human gtex_v8 context rif,species,group
# I thought we'd add some identifiers like
Provide answer on `glioma human gtex_v8` with context `[rif, species,group]`
```

Comment (Alexanderlacuna) [reply gn2/wqflask/views.py]: 1) Yeah, I conducted some experiments on this and others. It reduces the amount of 
"No feedback response/text does not contain".... responses we get
 also provides  more information on the subject with context provided
 Ideally Fahamu should be able to pick few keywords and get a response
 but currently it doesn't If this improves the normal search also, that would be okay
2) I would need to test this.

File: gn2/wqflask/views.py (hunk 1 above)
Comment (Alexanderlacuna) [reply gn2/wqflask/views.py]: These conditions are handled where token is  either

```
XAPIAN OPERATOR .e.g AND
RANGE 
```
More on this:
From  the docs this is the range syntax is word:
 https://xapian.org/docs/queryparser.html. 
This means that they are only considered ranges if they include the range operator (`..`) as part of their token. For example, `mean:5...7` should be rejected, whereas `mean` should be accepted as a word. 
Ideally this check should be  token has : and also .. and strip the word too
I will improve on this for clarity.

**Comment by zsloan** [comment]
Does it make sense to have the clean_xapian_query function in views.py? I think the idea with that file is just to contain the various routings for GN2.

That being said, I'm not sure where else would be a better place to put it, so this isn't a big deal. It's just that the file in question is already pretty long.

**Comment by Alexanderlacuna** [comment]
> Does it make sense to have the clean_xapian_query function in views.py? I think the idea with that file is just to contain the various routings for GN2.
> 
> That being said, I'm not sure where else would be a better place to put it, so this isn't a big deal. It's just that the file in question is already pretty long.

I could define it  inside the `/gnqna` endpoint since it is more of a utility for one of its inputs.

File: gn2/wqflask/views.py (hunk 1 above)
Comment (jnduli) [reply gn2/wqflask/views.py]: ah, I see. Thanks for the explanation. Then can we change this to:

```
prefix, _, suffix = token.partition(":") # line already defined below
if ".." in suffix and prefix in range_prefixes:
  continue
```

**Review by jnduli — APPROVED** [review]

**Pull request merged**